internal/olm/operator/registry/operator_installer.go

```diff
@@ -226,3 +233,53 @@ func (o OperatorInstaller) getInstalledCSV(ctx context.Context) (*v1alpha1.Clust
 	}
 	return csv, nil
 }
+
+// approveInstallPlan approves the install plan for a subscription, which will
+// generate a CSV
+func (o OperatorInstaller) approveInstallPlan(ctx context.Context, sub *v1alpha1.Subscription) error {
+	ip := v1alpha1.InstallPlan{}
+
+	ipKey := types.NamespacedName{
+		Name:      sub.Status.InstallPlanRef.Name,
+		Namespace: sub.Status.InstallPlanRef.Namespace,
+	}
+
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		if err := o.cfg.Client.Get(ctx, ipKey, &ip); err != nil {
+			return fmt.Errorf("error in getting install plan: %v", err)
+		}
+		// approve the install plan by setting Approved to true
+		ip.Spec.Approved = true
+		if err := o.cfg.Client.Update(ctx, &ip); err != nil {
+			return fmt.Errorf("error in approving install plan: %v", err)
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// waitForInstallPlan verifies if an Install Plan exists through subscription status
+func (o OperatorInstaller) waitForInstallPlan(ctx context.Context, sub *v1alpha1.Subscription) error {
+	subKey := types.NamespacedName{
+		Namespace: sub.GetNamespace(),
+		Name:      sub.GetName(),
+	}
+
+	ipCheck := wait.ConditionFunc(func() (done bool, err error) {
+		if err := o.cfg.Client.Get(ctx, subKey, sub); err != nil {
+			return false, err
+		}
+		if sub.Status.InstallPlanRef != nil {
+			return true, nil
+		}
+		return false, nil
+	})
+
+	if err := wait.PollImmediateUntil(200*time.Millisecond, ipCheck, ctx.Done()); err != nil {
+		return fmt.Errorf("install plan is not available for the subscription %s: %v", sub.Name, err)
+	}
+	return nil
+}
```

Comment:
Non-blocking: In one of these install plan functions, do we need to do a sanity check and ensure the install plan is for the CSV that we created in the subscription.

In the case of an upgrade being available, there can be multiple install plans and I think the subscription may reference the next available.

It may not be strictly necessary based in this case since we would need to approve the initial install plan to do the initial install and let OLM get to the point of doing an upgrade, BUT if it's a simple extra sanity check, it may be a good idea to go ahead and add.

Reply:
Great idea! I guess we can add the below check:

	if ip.Status.Plan != nil {
			for _, installPlan := range ip.Status.Plan {
				if installPlan.Resource.Kind == "ClusterServiceVersion" {
					if installPlan.Resolving != sub.Status.CurrentCSV {
						return fmt.Errorf("install plan CSV doesn't match the CSV in the subscription")
					}
				}
			}
		}